Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,1486 @@
+Authors: María Inés de Frutos Fernández, Xavier Généreux
+-/
+import Mathlib.Algebra.Algebra.Equiv
+import Mathlib.Algebra.Algebra.NonUnitalHom
+import Mathlib.Algebra.Module.BigOperators
+import Mathlib.Data.Finsupp.Basic
+
+/-!
+
+This file presents a skewed version of `Mathlib.Algebra.MonoidAlgebra.Basic`.
+
+We define `SkewMonoidAlgebra k G` attached with a skewed convolution product.
+Here, the product of two elements `f g : SkewMonoidAlgebra k G` is the finitely supported
+function whose value at `a` is the sum of `f x * (x • g y)` over all pairs `x, y`
+such that `x * y = a`.
+
+The associativity of the skewed multiplication depends on the `[MulSemiringAction G k]` instance.
+In particular, this means that unlike in `Mathlib.Algebra.MonoidAlgebra.Basic`, `G` will
+need to be a monoid for most of our uses.
+-/
+
+set_option autoImplicit false
+
+noncomputable section
+
+section
+
+/-- The skew monoid algebra over a semiring `k` generated by the monoid `G`.
+It is the type of finite formal `k`-linear combinations of terms of `G`,
+endowed with a skewed convolution product.
+-/
+@[ext]
+structure SkewMonoidAlgebra (k : Type*) (G : Type*) [Zero k] where ofFinsupp ::
+  /-- Map to the associated finitely supported function `G →₀ k`. -/
+  toFinsupp : G →₀ k
+
+open BigOperators
+open Function hiding Commute
+namespace SkewMonoidAlgebra
+
+universe u₁ u₂
+
+variable {k : Type u₁} {G : Type u₂} {H : Type*} {R : Type*}
+
+section AddCommMonoid
+
+variable [AddCommMonoid k]
+
+@[simp]
+theorem eta (f : (SkewMonoidAlgebra k G)) : ofFinsupp f.toFinsupp = f := by cases f; rfl
+
+private irreducible_def add :
+    (SkewMonoidAlgebra k G) → (SkewMonoidAlgebra k G) → (SkewMonoidAlgebra k G)
+  | ⟨a⟩, ⟨b⟩ => ⟨a + b⟩
+
+instance zero : Zero (SkewMonoidAlgebra k G) := ⟨⟨0⟩⟩
+
+instance add' : Add (SkewMonoidAlgebra k G) := ⟨add⟩
+
+instance smulZeroClass {S : Type*} [SMulZeroClass S k] :
+    SMulZeroClass S (SkewMonoidAlgebra k G) where
+  smul r p := ⟨r • p.toFinsupp⟩
+  smul_zero a := congr_arg ofFinsupp (smul_zero a)
+
+@[simp]
+theorem ofFinsupp_zero : (⟨0⟩ : SkewMonoidAlgebra k G) = 0 := rfl
+
+@[simp]
+theorem ofFinsupp_add {a b} : (⟨a + b⟩ : SkewMonoidAlgebra k G) = ⟨a⟩ + ⟨b⟩ :=
+  show _ = add _ _ by rw [add_def]
+
+@[simp]
+theorem ofFinsupp_smul {S : Type*} [SMulZeroClass S k] (a : S) (b) :
+    (⟨a • b⟩ : SkewMonoidAlgebra k G) = (a • ⟨b⟩ : SkewMonoidAlgebra k G) :=
+  rfl
+
+@[simp]
+theorem toFinsupp_zero : (0 : SkewMonoidAlgebra k G).toFinsupp = 0 := rfl
+
+@[simp]
+theorem toFinsupp_add (a b : SkewMonoidAlgebra k G) :
+    (a + b).toFinsupp = a.toFinsupp + b.toFinsupp := by
+  cases a
+  cases b
+  rw [← ofFinsupp_add]
+
+@[simp]
+theorem toFinsupp_smul {S : Type*} [SMulZeroClass S k] (a : S) (b : SkewMonoidAlgebra k G) :
+    (a • b).toFinsupp = a • b.toFinsupp :=
+  rfl
+
+theorem _root_.IsSMulRegular.skewMonoidAlgebra {S : Type*} [Monoid S] [DistribMulAction S k] {a : S}
```

Comment:
This lemma could be deduced from the fact that the skewMonoidAlgebra is a free k-module. Is it worth doing it here?

Reply:
I don't have a strong opinion on the lemma's position; it follows the structure used in the Polynomial file.

We don't currently have the fact that a SkewMonoidAlgebra is a free k-module, but I can add it if you think it's relevant.
As for the fact that the lemma entails from the free module structure, we chose this proof because it matches the idea from `finsupp` and `polynomial`. I think it's more natural to use the `Finsupp` structure than to do a small detour with the k-module structure. What do you think?

Reply:
The freeness of the `SkewMonoidAlgebra` is important to add here, because it follows from the construction, but not from the universal property.

Reply:
Would you be ok with me adding the freeness independently of this? Meaning I would leave `IsSMulRegular.skewMonoidAlgebra` as is.

An other problem is that the file is getting really long.

Reply:
The freeness is proved below, as a direct application of [Module.Free.of_equiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/FreeModule/Basic.html#Module.Free.of_equiv).
Two lemmas are apparently missing, 1/ that freeness implies the IsSMulRegular condition, 2/ that this condition is preserved under a LinearEquiv.

Reply:
Will do